### PR TITLE
PXB-2242 Prevent back up if Source system (DB) version is higher the…

### PR DESCRIPTION
**Pull request by rahulmalik87**
…n current PXB version

Add a configuration parameter --no-server-version-check to allow override
of the check process.  The default is override is off (must turn on override) 
If same base upstream/Percona version and PXB version - allow back up to proceed
If base upsteam/Percona version is less then PXB version - allow back up
to proceed
If base upstream/Percona version is greater (newer) then PXB version and
no-server-version-check is  set to allow backup - allow back up to proceed 
If base upstream/Percona version is greater (newer) then PXB version and
no-server-version-check is NOT set to allow back up - prevent back from
occurring - provide compete notification

### CODE REVIEW

**Comment by rahulmalik87** [comment]
jenkin results https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param/lastCompletedBuild/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 1]
```diff
@@ -407,6 +407,23 @@ static bool check_server_version(unsigned long version_number,
     }
   }
 
+  std::string mysql_version = MYSQL_SERVER_VERSION;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: I would change the variable name to be more clear of what it is about. `mysql_version` seems like this is the server version we are trying to backup, while in fact this is pxb base mysql version.

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 2]
```diff
@@ -407,6 +407,23 @@ static bool check_server_version(unsigned long version_number,
...
+
+  // replace 8.0.21 first . with 0 and second . with null
+  std::size_t found = mysql_version.find_first_of(".");
+  if (found != std::string::npos) mysql_version[found] = '0';
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: This assumes PXB will always have x.x.xx version pattern, which will require to re-work this function in future. For example, I did a test with PS 8.10.22 and PXB 8.10.21:
```
(gdb) p mysql_version_int
+p mysql_version_int
$3 = 801021
(gdb) p version_number
+p version_number
$4 = 81022
```
I would make this function to use the same logic as mysql_get_server_version, which is where `version_number` is extracted from.

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 3]
```diff
@@ -407,6 +407,23 @@ static bool check_server_version(unsigned long version_number,
...
+  found = mysql_version.find_first_of(".");
+  if (found != std::string::npos)
+    mysql_version.erase(mysql_version.begin() + found);
+
+  unsigned long mysql_version_int = stoi(mysql_version);
+  if (!opt_no_server_version_check && mysql_version_int < version_number) {
+    msg("Error: Unspported server version %s, Use latest version of PXB "
+        "or to continue with risk, use --no-server-version-check",
+        MYSQL_SERVER_VERSION);
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: > Error: Unspported server version %s

This is reporting PXB version (`MYSQL_SERVER_VERSION`), while it should be reporting MySQL server version - `version_number`.

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]

**Comment by altmannmarcelo** [comment]
LGTM

**Review by altmannmarcelo — APPROVED** [review]

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 4]
```diff
@@ -363,6 +364,25 @@ void parse_show_engine_innodb_status(MYSQL *connection) {
...
+    minor = stoi(pxb_base.substr(major_p + 1, minor_p - major_p));
+
+  std::size_t version_p = pxb_base.find(".", minor_p + 1);
+  if (version_p != std::string::npos)
+    version = stoi(pxb_base.substr(minor_p + 1, version_p - minor_p));
+  else
+    version = stoi(pxb_base.substr(minor_p + 1));
+  return major * 10000 + minor * 100 + version;
+}
+
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_mysql.cc:385]: Does it work for debug binaries? And if the server is debug version?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/xtrabackup.cc [hunk 5]
```diff
@@ -1020,6 +1022,13 @@ struct my_option xb_client_options[] = {
      (uchar *)&opt_force_non_empty_dirs, (uchar *)&opt_force_non_empty_dirs, 0,
      GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
 
+    {"no-server-version-check", OPT_NO_SERVER_VERSION_CHECK,
+     "This option allow backup to proceed"
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xtrabackup.cc]:
```suggestion
     "This option allows backup to proceed"
```

File: storage/innobase/xtrabackup/test/t/version_check.sh [hunk 6]
```diff
@@ -10,6 +10,10 @@
 
 start_server
 
+#test server version check
+xtrabackup --backup --no-server-version-check --target-dir=$topdir/backup1
+rm -r $topdir/backup1
+
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/version_check.sh:16]: Can you add a test case for a negative scenario?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 4 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_mysql.cc:385]: yes, it works for debug version,  release version and patch version. 
i.e it works for all version like 8.0.12 or  8.0.12.debug or 8.0.12.1

File: storage/innobase/xtrabackup/test/t/version_check.sh (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/t/version_check.sh:16]: If you find it hard to introduce have_debug.sh/inc, please let me know.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Comment by satya-bodapati** [comment]
Also click on re-request review, when it is ready for review again

**Comment by rahulmalik87** [comment]
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param/lastBuild/

mdl_lock.sh  is due to --lock-ddl-per-table
xb_parallel_encrypt.sh  is successful on local. 
will give another round to see if it is succesful.

File: storage/innobase/xtrabackup/src/xtrabackup.cc (hunk 5 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/xtrabackup.cc]: Why do I still see "allow" instead of "allows?" I see that you marked it as resolved..

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Review by altmannmarcelo — APPROVED** [review]
LGTM

**Review by satya-bodapati — APPROVED** [review]
Please squash before merging
